### ISSUE TITLE
game runs, but connection doesn't work

### ISSUE THREAD

**Issue by RonXTCdaBass** [erodozer/twitch.idlegame#1]
hello at first ^^

maybe a nice idea, but it seems the connection method isnt working. 
i tried everything for a few hours. :(

```
07:31:23.893 XHRPOST   https://id.twitch.tv/oauth2/token   [HTTP/2 400 Bad Request 186ms]

07:31:24.116 [07:31] info: Sending authentication to server.. tmi.js:2794:15

07:31:24.305 [07:31] error: Login authentication failed tmi.js:2794:15

```

a better description and or a complete readme file would nice.
maybe possible changing the login to standart OAUTH ??

**Comment by RonXTCdaBass** [comment]
i don't know if it´s okay, but i want to show you my (still NOT working) idea...
at the moment, i doesn't have the knowledge to complete this until it works...

![TwitchIdleGame_idea_screen](https://github.com/user-attachments/assets/5b92bacb-1bcb-45e3-8eb2-c74a05ed89f9)


```
<!-- twitch connection -->
<script src="https://github.com/tmijs/tmi.js/releases/download/v1.8.5/tmi.js"></script>
        <script>
		window.onload = function() {
	i_user = document.getElementById("i_user")
    i_oauth = document.getElementById("i_oauth")
    i_target = document.getElementById("i_target")
	o_status = document.getElementById("o_status")
		}
		
        var u_user
        var u_oauth
		var u_target

		var client = new tmi.Client({
            channels: []
        });
		
		
		function connection() {
            update_client();
            console.log("Starting connection")
            o_status.innerHTML = "Connected and waiting..."
            
			client.connect()
                .catch(
                    err => {
						console.error(err)
                        console.log("Connection ERROR");
						localStorage.removeItem('idle_game.twitch_token');
                    });
        }

        function disconnection() {
            console.log("Starting disconnection")
            client.disconnect()
			o_status.innerHTML = "Disconnected..."
            
        }
		
		
		function update_client() {
			u_target = i_target.value
            u_user = i_user.value
            u_oauth = i_oauth.value
            client = new tmi.Client({
                options: { debug: true },
				channels: [ config.twitch.channel ],
                identity: {
                    username: u_user,
                    password: u_oauth
                }
						
            });
			console.log(`login try ${u_user} `);

		localStorage.setItem('idle_game.twitch.clientId', u_user);  //???
		localStorage.setItem('idle_game.config.twitch.channel', u_target);  //???
		localStorage.setItem('idle_game.twitch_token', u_oauth);  //???
                return `oauth:${u_oauth}`;
        }

```


and the html part before the closing body tag...    



```
<hr>
		<div class="twitch_connect"style="height:2em"><p>
		<span style="float:left" class="user-select-none "><b>Status: </b><code id="o_status">...</code></span>
		<span style="float:right"><button onclick="connection()"><font color="green">✔</font>Connect</button>
							<button onclick="disconnection()"><font color="red">❌</font>Disconnect</button></span>
		</p></div>
	<hr>
		<div class="settings">
            <h3 class="user-select-none" style="text-align:center">Connection Settings</h3>
                <div class="form">
					<div class="input-group">
						<span class="input-group-text"style="color:red">target channel</span> @ twitch.tv/
                        <input id="i_target" type="text" class="form-control fs_input_todisable" placeholder="channel">
					</div>
					<div class="input-group">
						<span class="input-group-text"style="color:orange">user/bot-name</span> @ twitch.tv/
                        <input id="i_user" type="text" class="form-control fs_input_todisable" placeholder="username">
					</div>
					<div class="input-group">
                        <span class="input-group-text"style="color:darkgreen">Your oauth token: </span>
						<a class="btn-outline-info" href="https://dev.twitch.tv/docs/authentication/" target=”_blank”>(?)</a>
						<a class="btn-outline-info" href="https://twitchapps.com/tmi/" target=”_blank”>Get !</a>
                        <input id="i_oauth" type="password" class="form-control fs_input_todisable" placeholder="OAUTH:XXXXXXXX...">
                    </div>
                </div>
		</div>
```

**Comment by erodozer** [comment]
Hi, this project was just something I spent a single stream putting together.  I did it to demonstrate how to make something quickly for chat interaction with twitch + OBS.  I do not intend to make much out of it or accept changes to it in its current state.  If I revisit it, it'll be to rewrite it into something production quality instead of a simple jam project.

Bot connection won't work out of the box because the git repo does not include the required file for authentication, it's purposely ignored by git because it contains sensitive information.  It is only required if you wish to have a bot that can write to the target channel.  You need to create a `secrets.js` file exactly as described in the readme, with your own sensitive values provided.  The access token should be for the bot user.  You can make one using the twitch cli.

**Comment by RonXTCdaBass** [comment]
Thanks for answering but

Sorry i'm not stupid.
I specifically registered a new app, took new data (app ID, token, secret) and created the aforementioned Secrets.js.

the console said:


`14:00:39.550 XHRPOST   https://id.twitch.tv/oauth2/token   [HTTP/2 403 Forbidden 249ms]`


is there another file or trick needed ?

Unfortunately, for me the game is useless if it doesn't respond to chat.

**Comment by erodozer** [comment]
When you get a 403 from the oauth api the request body will typically state the error.  In the case of a 403, it's usually some field is missing or has an invalid value.

When getting the token from the CLI, irc bots need certain scopes to be able to read and write.
```
twitch token -u -s "chat:read chat:edit"
```
This will give you an access token and refresh token, both of which are required with the way this project set up its auth.  If you're still having trouble after setting all those values in your `secrets.js`, try clearing the localStorage, as the project is naively keeping the new refreshToken there after login.

It's an old project, I've learned to do things better when dealing with the twitch API since I hacked this out, I made this whole project in about 6 hours.  I did notice that support for bot users is incomplete, the username is assumed to be the channel name that's in config.js instead of the username in secrets.  That's a minor change on game.html#L946.  

However, I can confirm that if using the same account for channel and user, ie assuming your streamer account as a bot, this is working without modification.  I loaded up a fresh copy of the project this morning with new credentials and no other modifications.

![image](https://github.com/user-attachments/assets/f7ecafac-08de-4990-bcf4-c6e7bff5144b)
